### PR TITLE
fix(types): add sx prop whenever className can be string

### PR DESCRIPTION
**Pull request by hasparus**
I'm not exactly sure if we have the best possible implementation, as we don't produce `className` at runtime, we produce `css` and call Emotion's `jsx`, so it seems that we should also use their types for correctness.

cc @atanasster @dcastil

Relevant to #1307, but I'm not sure if it closes.

---

There are three changes here.

> - **breaking TypeScript**: Renamed and removed types.
>   - `SxProps` to `SxProp`.
>   - `SxStyleProp`, an alias for `ThemeUIStyleObject` removed. Use `ThemeUIStyleObject` instead.
> - Fix: Add `sx` props types to all props accepting `className`.
> - Fix WithPoorAsProp to work with ComponentProps utility type.

The change to `WithPoorAsProp` allows to read props of `@theme-ui/mdx` `Styled` on type level. Previously, after passing throught ComponentProps, they became dicts of `unknown`.

```tsx
type MyProps = ComponentPropsWithoutRef<typeof Styled.p>
```

The change to `WithConditionalSxProp` allows to use `sx` with all components accepting `className` of type `string`, even if they also accept other values for `className`.

### CODE REVIEW

File: packages/core/src/jsx-namespace.ts [hunk 1]
```diff
@@ -1,10 +1,8 @@
-import { SxProps } from './types'
+import { SxProp } from './types'
 
-type WithConditionalSxProps<P> = 'className' extends keyof P
-  ? P extends { className?: string }
-    ? Omit<P, keyof SxProps> & SxProps
-    : Omit<P, keyof SxProps>
-  : Omit<P, keyof SxProps>
+export type WithConditionalSxProp<P> = { className: string } extends P
```
Comment (dcastil) [inline packages/core/src/jsx-namespace.ts:3]: I like this one much more than the previous solution. I don't know why Emotion did it that way.

**Comment by hasparus** [comment]
Okay, I got greedy, and there's going to be another change, what you can probably see in failed tests. I'm making `@theme-ui/core` and `theme-ui` strict TypeScript.

File: packages/core/src/jsx-namespace.ts (hunk 1 above)
Comment (hasparus) [reply packages/core/src/jsx-namespace.ts]: Sorry, Dany. Now it's even uglier. I forgot that components can have more than one prop while writing this, and that's how I got that nice one-liner :D

Comment (dcastil) [reply packages/core/src/jsx-namespace.ts]: haha 😂

**Comment by hasparus** [comment]
I didn't even touch this :cry: 
image

LibraryManagedAttributes can literally wreak generic components inference :D Okay, this needs to be fixed on our side. I don't want to add "our jsx pragma is incompatible with react-hook-form" to the readme :D 

image

**Comment by hasparus** [comment]
cc @beerose

**Pull request merged**